jquery-promise.es6

```diff
@@ -0,0 +1,60 @@
+/**
+ * Miniature ES6 Promise polyfill based on jQuery.Deferred.
+ *
+ * jQuery 3 is Promises/A+ compliant, so this should be good enough for
+ * most situations in which jQuery is being imported anyway.
+ */
+
+import { Deferred, when, each } from 'jquery';
+
+function finallyImp(callback) {
+    return this.then(callback, callback);
+}
+
+function toPromise(deferred, target) {
+    const promise = deferred.promise(target);
+    promise.finally = finallyImp;
+    return promise;
+}
+
+export default function Promise(executor) {
+    const deferred = Deferred();
+    const resolve = deferred.resolve.bind(deferred),
+        reject = deferred.reject.bind(deferred);
+    executor(resolve, reject);
+    return toPromise(deferred);
+}
+
+Promise.all = function(iterable) {
+    return toPromise(when.apply(null, iterable));
+}
+
+Promise.race = function(iterable) {
+    const deferred = Deferred();
+    function resolved(value) {
+        if (deferred.state() === 'pending') deferred.resolve(value);
+    }
+    function rejected(reason) {
+        if (deferred.state() === 'pending') deferred.reject(reason);
+    }
+    each(iterable, (key, promise) => promise.then(resolved, rejected));
+    return toPromise(deferred);
+}
+
+Promise.resolve = function(value) {
+    return new Promise((resolve, reject) => {
+        if (value.then && value.then instanceof Function) {
+            value.then(resolve, reject);
+        } else {
+            resolve(value);
+        }
+    });
+}
+
+Promise.reject = function(reason) {
+    return new Promise((resolve, reject) => reject(reason));
+}
+
+const local = global || self || window || Function('return this')();
+
+if (local && !local.Promise) local.Promise = Promise;
```

Comment:
This `if` could be done at the beginning, [most browsers support `Promise` natively](https://caniuse.com/#feat=promises) they won't need this code to run.

Reply:
I considered this, but it would become a problem when I want to import the Promise implementation into my unittest module (there would only be a Promise implementation to export in environments that don't have a native Promise). It's only a very small amount of code and the browser is still going to parse it even if you slap a big if-statement around it, so I don't feel very sorry for the unneeded function definitions.